src/stream/src/executor/aggregation/state_cache/mod.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+use risingwave_common::array::Op;
+use risingwave_common::types::{Datum, DatumRef};
+use smallvec::SmallVec;
+
+use self::cache::OrderedCache;
+use super::minput::StateCacheInputBatch;
+
+pub mod array_agg;
+mod cache;
+pub mod extreme;
+pub mod string_agg;
+
+/// Cache key type.
+pub type CacheKey = Vec<u8>;
+
+/// Trait that defines the interface of state table cache.
+pub trait StateCache: Send + Sync + 'static {
+    /// Check if the cache is synced with state table.
+    fn is_synced(&self) -> bool;
+
+    /// Apply a batch of updates to the cache.
+    fn apply_batch(&mut self, batch: StateCacheInputBatch<'_>);
+
+    /// Begin syncing the cache with state table.
+    fn begin_syncing(&mut self) -> StateCacheFiller<'_>;
+
+    /// Get the aggregation output.
+    fn get_output(&self) -> Datum;
+}
+
+/// Cache maintenance interface.
+/// Note that this trait must be private, so that only [`StateCacheFiller`] can use it.
+trait StateCacheMaintain: Send + Sync + 'static {
+    /// Insert an entry to the cache without checking row count, capacity, key order, etc.
+    /// Just insert into the inner cache structure, e.g. `BTreeMap`.
+    fn insert_unchecked(&mut self, key: CacheKey, value: SmallVec<[DatumRef<'_>; 2]>);
+
+    /// Mark the cache as synced.
+    fn set_synced(&mut self);
+}
+
+/// A temporary handle for filling the state cache.
+/// The state cache will be marked as synced automatically when this handle is dropped.
+pub struct StateCacheFiller<'a> {
+    capacity: usize,
+    cache: &'a mut dyn StateCacheMaintain,
```

Comment:
Then we will have dynamic dispatch here, each row. :(

Reply:
Would you like to test whether collecting a vector of `(CacheKey, SmallVec)` and `insert` it in a batch will perform better? 🤔 ... or even taking a `impl Stream`?

Reply:
> Then we will have dynamic dispatch here, each row. :(

I actually noticed this, but I thought the overhead of dynamic dispatch here is ignorable compared to state table reads which incurs IO.

Reply:
Oh, I see. This will only occur when the cache is completely invalidated, right? Then there'll be no much problem.